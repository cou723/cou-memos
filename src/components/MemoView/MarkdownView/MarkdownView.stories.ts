import type { Meta, StoryObj } from "@storybook/react";

import { MarkdownView } from ".";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
    title: "MarkdownView",
    component: MarkdownView,
    tags: ["autodocs"]
} satisfies Meta<typeof MarkdownView>;

export default meta;
type Story = StoryObj<typeof meta>;

// headings
export const H1: Story = { args: { text: "# Hello" } };
export const H2: Story = { args: { text: "## Hello" } };
export const H3: Story = { args: { text: "### Hello" } };

export const List: Story = {
    args: {
        text: `- list1
- list2
- list3`
    }
};

export const Paragraph: Story = {
    args: { text: `#tag1 text` }
};

export const InlineCode: Story = {
    args: { text: `\`Hello\`` }
};

export const All: Story = {
    args: {
        text: `#tag1 text
# heading 1
## heading 2
### heading 3
- list1
- list2
- list3

\`code\`

\`\`\`js
console.log();
\`\`\`

\`\`\`
normal code
\`\`\`
`
    }
};

export const ConsecutiveElements: Story = {
    args: {
        text: `p1 \`code\` p2`
    }
};
