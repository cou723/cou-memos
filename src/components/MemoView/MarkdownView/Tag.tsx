import React from "react";
import { Badge } from "react-daisyui";

export const Tag = React.memo(({ text }: { text: string }) => {
    return <Badge>{text}</Badge>;
});
