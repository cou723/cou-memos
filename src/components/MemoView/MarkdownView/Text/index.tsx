import type React from "react";

type Props = { children: React.ReactNode[] };

export const Text = ({ children }: Props) => {
    return <div>{children.map((element) => element)}</div>;
};
