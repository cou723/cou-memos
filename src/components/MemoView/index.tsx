import type { FC } from "react";
import React from "react";

import { Card } from "react-daisyui";

import { Actions } from "./Actions";
import { MarkdownView } from "./MarkdownView";

import type { Memo } from "@/types/memo";

type Props = {
    memo: Memo;
    onEdit: (id: number) => void;
    onDelete: (id: number) => void;
};

// eslint-disable-next-line react/display-name
export const MemoView: FC<Props> = React.memo(
    ({ memo, onEdit, onDelete }: Props) => {
        const handleEdit = () => onEdit(memo.id);
        const handleDelete = () => onDelete(memo.id);

        return (
            <>
                <Card>
                    <Card.Body className="p-3">
                        <MarkdownView text={memo.text} />
                        <Actions
                            handleDelete={handleDelete}
                            handleEdit={handleEdit}
                            memo={memo}
                        />
                    </Card.Body>
                </Card>
            </>
        );
    },
);
