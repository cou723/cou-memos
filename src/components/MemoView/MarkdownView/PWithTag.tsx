import React, { FC } from "react";
import { Tag } from "./Tag";
import { splitTags } from "@/lib/extractTag";

function isTag(text: string): boolean {
    return text.startsWith("#");
}

type Props = { text: string };

export const PWithTag: FC<Props> = React.memo(({ text }) => {
    const elements: JSX.Element[] = [];

    const texts = splitTags(text);

    for (const [index, text] of texts.entries()) {
        if (isTag(text.text)) elements.push(<Tag text={text.text} key={index} />);
        else {
            elements.push(
                <>
                    {text.before != null ? text.before == "\n" ? <br /> : text.before : ""}
                    {text.text}
                </>
            );
        }
    }

    return <>{elements.map((element) => element)}</>;
});
