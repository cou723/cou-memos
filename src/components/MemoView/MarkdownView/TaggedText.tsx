import React, { FC } from "react";
import { Tag } from "./Tag";
import { splitTags } from "@/lib/extractTag";

function isTag(text: string): boolean {
    return text.startsWith("#");
}

type Props = { text: string };

export const TaggedText: FC<Props> = React.memo(({ text }) => {
    const elements: JSX.Element[] = [];

    const texts = splitTags(text);

    for (const [index, text] of texts.entries()) {
        if (isTag(text.text)) {
            console.log("set key:", index);
            elements.push(<Tag text={text.text} key={index} />);
        } else {
            elements.push(
                <p key={index + text.text}>
                    {text.before != null ? text.before == "\n" ? <br /> : text.before : ""}
                    {text.text}
                </p>
            );
        }
    }

    return <>{elements.map((element) => element)}</>;
});
