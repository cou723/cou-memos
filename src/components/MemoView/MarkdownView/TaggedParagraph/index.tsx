import React, { FC } from "react";
import { Tag } from "../Text/Tag";
import { splitTags } from "@/lib/extractTag";

function isTag(text: string): boolean {
    return text.startsWith("#");
}

type Props = { text: string };

export const TaggedParagraph: FC<Props> = React.memo(({ text }) => {
    const elements: JSX.Element[] = [];

    const texts = splitTags(text);

    for (const [index, content] of texts.entries()) {
        if (isTag(content.text)) {
            elements.push(<Tag text={content.text} key={index} />);
        } else {
            elements.push(
                <>{(content.before != null ? content.before == "\n" ? <br /> : content.before : "") + content.text}</>
            );
        }
    }

    return <p style={{}}>{elements.map((element) => element)}</p>;
});
