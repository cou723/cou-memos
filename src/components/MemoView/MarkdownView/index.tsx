import React from "react";
import ReactMarkdown from "react-markdown";
import { Prism as SyntaxHighlighter } from "react-syntax-highlighter";
import { PWithTag } from "./PWithTag";

export const MarkdownView = React.memo(({ text }: { text: string }) => {
    return (
        <ReactMarkdown
            components={{
                code({ node, inline, className, children, ...props }) {
                    const match = /language-(\w+)/.exec(className || "");
                    return !inline && match ? (
                        <SyntaxHighlighter
                            {...props}
                            children={String(children).replace(/\n$/, "")}
                            language={match[1]}
                            PreTag="div"
                        />
                    ) : (
                        <code {...props} className={className}>
                            {children}
                        </code>
                    );
                },
                p({ node, children, ...props }) {
                    return <PWithTag text={children.toString()} />;
                }
            }}
            className="memo"
        >
            {text}
        </ReactMarkdown>
    );
});
