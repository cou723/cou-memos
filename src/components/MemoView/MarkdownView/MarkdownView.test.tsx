/* eslint-disable @typescript-eslint/no-floating-promises */
// @vitest-environment jsdom

import { cleanup, render } from "@testing-library/react";
import { describe, afterEach, it, expect } from "vitest";
import { MarkdownView } from "@/components/MemoView/MarkdownView";
import {
    All,
    ConsecutiveElements,
    H1,
    H2,
    H3,
    InlineCode,
    List,
    Paragraph
} from "@/components/MemoView/MarkdownView/MarkdownView.stories";

describe("MarkdownView single element", () => {
    afterEach(() => {
        cleanup();
    });

    it("h1", async () => {
        const { asFragment, getByText } = render(<MarkdownView {...H1.args} />);
        expect(asFragment()).toMatchSnapshot();

        expect(getByText("Hello").tagName).equal("H1");
    });

    it("h2", async () => {
        const { asFragment, getByText } = render(<MarkdownView {...H2.args} />);
        expect(asFragment()).toMatchSnapshot();

        expect(getByText("Hello").tagName).equal("H2");
    });

    it("h3", async () => {
        const { asFragment, getByText } = render(<MarkdownView {...H3.args} />);
        expect(asFragment()).toMatchSnapshot();

        expect(getByText("Hello").tagName).equal("H3");
    });

    it("p", async () => {
        const { asFragment, getByText } = render(<MarkdownView {...Paragraph.args} />);
        expect(asFragment()).toMatchSnapshot();

        expect(getByText("#tag1").tagName).equal("DIV");
        expect(getByText("text").tagName).equal("P");
    });

    it("list", async () => {
        const { asFragment, getByText } = render(<MarkdownView {...List.args} />);
        expect(asFragment()).toMatchSnapshot();

        expect(getByText("list1").tagName).equal("LI");
        expect(getByText("list2").tagName).equal("LI");
        expect(getByText("list3").tagName).equal("LI");
    });

    it("inline code", async () => {
        const { asFragment, getByText } = render(<MarkdownView {...InlineCode.args} />);
        expect(asFragment()).toMatchSnapshot();
        expect(getByText("Hello").tagName).equal("CODE");
    });

    it("code", async () => {
        const { asFragment, getByText } = render(
            <MarkdownView
                text={`\`\`\`c
#include
int main(){
    printf("hello world");
}

\`\`\``}
            />
        );
        expect(asFragment()).toMatchSnapshot();
        expect(getByText("printf").tagName).equal("SPAN");
    });
});

describe("MarkdownView multiple elements", () => {
    afterEach(() => {
        cleanup();
    });
    it("h1, h2, h3, p, list, inline code, code", async () => {
        const { asFragment, getByText, container } = render(<MarkdownView {...All.args} />);

        expect(asFragment()).toMatchSnapshot();

        expect(getByText("heading 1").tagName).equal("H1");
        expect(getByText("heading 2").tagName).equal("H2");
        expect(getByText("heading 3").tagName).equal("H3");
        expect(getByText("list1").tagName).equal("LI");

        expect(container.getElementsByTagName("code").length).equal(3);
        expect(container.getElementsByTagName("pre").length).equal(2);

        expect(getByText("console").tagName).equal("SPAN");
        expect(getByText(".").tagName).equal("SPAN");
        expect(getByText("log").tagName).equal("SPAN");
        expect(getByText("(").tagName).equal("SPAN");
        expect(getByText(")").tagName).equal("SPAN");
        expect(getByText(";").tagName).equal("SPAN");

        expect(getByText("normal code").tagName).equal("SPAN");
    });

    it("p, inline code, p", async () => {
        const { asFragment, getByText } = render(<MarkdownView {...ConsecutiveElements.args} />);

        expect(asFragment()).toMatchSnapshot();

        expect(getByText("p1").tagName).equal("P");
        expect(getByText("code").tagName).equal("CODE");
        expect(getByText("p2").tagName).equal("P");
    });
});
