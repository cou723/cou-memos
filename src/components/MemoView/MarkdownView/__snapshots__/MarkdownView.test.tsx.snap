// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MarkdownView multiple elements > h1, h2, h3, p, list, inline code, code 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <div
      aria-label="Badge"
      class="badge mr-2 block badge-neutral"
    >
      #tag1
    </div>
    <p>
      text
    </p>
    

    <h1>
      heading 1
    </h1>
    

    <h2>
      heading 2
    </h2>
    

    <h3>
      heading 3
    </h3>
    

    <ul>
      

      <li>
        list1
      </li>
      

      <li>
        list2
      </li>
      

      <li>
        list3
      </li>
      

    </ul>
    

    <code
      class="inline-code"
      node="[object Object]"
    >
      code
    </code>
    

    <pre>
      <div
        node="[object Object]"
        style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none; padding: 1em; margin: .5em 0px; overflow: auto; border-radius: 0.3em; background: rgb(29, 31, 33);"
      >
        <code
          class="language-js"
          style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none;"
        >
          <span
            class="token console"
            style="color: rgb(255, 255, 182); text-decoration: underline;"
          >
            console
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            .
          </span>
          <span
            class="token method property-access"
            style="color: rgb(218, 208, 133);"
          >
            log
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            (
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            )
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ;
          </span>
        </code>
      </div>
    </pre>
    

    <pre>
      <div
        node="[object Object]"
        style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none; padding: 1em; margin: .5em 0px; overflow: auto; border-radius: 0.3em; background: rgb(29, 31, 33);"
      >
        <code
          style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none;"
        >
          <span>
            normal code
          </span>
        </code>
      </div>
    </pre>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView multiple elements > p, inline code, p 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <p>
      p1 
    </p>
    <code
      class="inline-code"
      node="[object Object]"
    >
      code
    </code>
    <p>
       p2
    </p>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView single element > code 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <pre>
      <div
        node="[object Object]"
        style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none; padding: 1em; margin: .5em 0px; overflow: auto; border-radius: 0.3em; background: rgb(29, 31, 33);"
      >
        <code
          class="language-c"
          style="color: rgb(197, 200, 198); text-shadow: 0 1px rgba(0, 0, 0, 0.3); font-family: Inconsolata, Monaco, Consolas, 'Courier New', Courier, monospace; direction: ltr; text-align: left; white-space: pre; word-spacing: normal; word-break: normal; line-height: 1.5; tab-size: 4; hyphens: none;"
        >
          <span
            class="token macro directive-hash"
            style="color: rgb(150, 203, 254);"
          >
            #
          </span>
          <span
            class="token macro directive"
            style="color: rgb(150, 203, 254);"
          >
            include
          </span>
          <span>
            

          </span>
          <span />
          <span
            class="token"
            style="color: rgb(150, 203, 254);"
          >
            int
          </span>
          <span>
             
          </span>
          <span
            class="token"
            style="color: rgb(218, 208, 133);"
          >
            main
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            (
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            )
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            {
          </span>
          <span>
            

          </span>
          <span>
                
          </span>
          <span
            class="token"
            style="color: rgb(218, 208, 133);"
          >
            printf
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            (
          </span>
          <span
            class="token"
            style="color: rgb(168, 255, 96);"
          >
            "hello world"
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            )
          </span>
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            ;
          </span>
          <span>
            

          </span>
          <span />
          <span
            class="token"
            style="color: rgb(197, 200, 198);"
          >
            }
          </span>
          <span>
            

          </span>
        </code>
      </div>
    </pre>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView single element > h1 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <h1>
      Hello
    </h1>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView single element > h2 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <h2>
      Hello
    </h2>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView single element > h3 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <h3>
      Hello
    </h3>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView single element > inline code 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <code
      class="inline-code"
      node="[object Object]"
    >
      Hello
    </code>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView single element > list 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <ul>
      

      <li>
        list1
      </li>
      

      <li>
        list2
      </li>
      

      <li>
        list3
      </li>
      

    </ul>
  </div>
</DocumentFragment>
`;

exports[`MarkdownView single element > p 1`] = `
<DocumentFragment>
  <div
    class="memo"
  >
    <div
      aria-label="Badge"
      class="badge mr-2 block badge-neutral"
    >
      #tag1
    </div>
    <p>
      text
    </p>
  </div>
</DocumentFragment>
`;
